metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -354,6 +357,25 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
             throw new BrokerIdNotRegisteredException("Controller is in pre-migration mode and cannot register KRaft " +
                 "brokers until the metadata migration is complete.");
         }
+
+        if (featureControl.metadataVersion().isDirectoryAssignmentSupported()) {
+            Set<Uuid> set = new HashSet<>(request.logDirs());
+            if (set.stream().anyMatch(DirectoryId::reserved)) {
+                throw new InvalidRequestException("Reserved directory ID in request");
+            }
+            if (set.size() != request.logDirs().size()) {
+                throw new InvalidRequestException("Duplicate directory ID in request");
+            }
+            for (BrokerRegistration registration : brokerRegistrations().values()) {
```

Comment:
Another thing. This is O(num_brokers * num_directories) which isn't good for larger deployments. I think we should just have a `TimelineHashMap<Uuid, Integer>` that maps log dir UUID to the broker that has it. The main hassle is removing the mappings during broker unregistration, but that is rare, of course.

Reply:
> Also, we shouldn't be checking the previous registration for this specific broker ID... it would be quite normal to re-register broker ID X with the same set of directories it had last time we registered it

This might be what's causing the test failures btw

Reply:
@cmccabe:

>  InvalidRequestException seems too generic

I've introduced InvalidRegistrationException to handle these.

> Also, we shouldn't be checking the previous registration for this specific broker ID.

I forgot that. Thanks for pointing this out!

> we should just have a TimelineHashMap<Uuid, Integer> that maps log dir UUID to the broker that has it

Great suggestion. Added.